test/tstDetailsRay.cpp

```diff
@@ -85,24 +86,40 @@ BOOST_AUTO_TEST_CASE(intersects_box)
   BOOST_TEST(intersects(Ray{{2, 2, .5}, {-1, -1, 0}}, unit_box));
   BOOST_TEST(intersects(Ray{{2, -1, .5}, {-1, 1, 0}}, unit_box));
   BOOST_TEST(intersects(Ray{{2, 1, .5}, {-1, -1, 0}}, unit_box));
-  BOOST_TEST(intersects(Ray{{0, 0, .5}, {1, 1, 0}}, unit_box));
-  BOOST_TEST(!intersects(Ray{{0, 0, .5}, {-1, -1, 0}}, unit_box));
-  BOOST_TEST(intersects(Ray{{1, 1, .5}, {-1, -1, 0}}, unit_box));
-  BOOST_TEST(!intersects(Ray{{1, 1, .5}, {1, 1, 0}}, unit_box));
 
   // hit a corner
   BOOST_TEST(intersects(Ray{{-1, 1.5, 1.5}, {1, -1, -1}}, unit_box));
   BOOST_TEST(!intersects(Ray{{-1, 1, 1}, {-1, 0, 0}}, unit_box));
   BOOST_TEST(!intersects(Ray{{-1, 1, 1}, {0, 1, 1}}, unit_box));
-  BOOST_TEST(intersects(Ray{{1, 1, 1}, {-1, -1, -1}}, unit_box));
-  BOOST_TEST(!intersects(Ray{{1, 1, 1}, {1, 1, 1}}, unit_box));
-  BOOST_TEST(intersects(Ray{{0, 0, 0}, {1, 1, 1}}, unit_box));
-  BOOST_TEST(!intersects(Ray{{0, 0, 0}, {-1, -1, -1}}, unit_box));
   BOOST_TEST(!intersects(Ray{{2, 1, 1}, {1, 0, 0}}, unit_box));
   BOOST_TEST(!intersects(Ray{{2, 1, 1}, {0, 1, 1}}, unit_box));
   BOOST_TEST(intersects(Ray{{-1, -1, -1}, {1, 1, 1}}, unit_box));
   BOOST_TEST(intersects(Ray{{2, 2, 2}, {-1, -1, -1}}, unit_box));
 
   BOOST_TEST(!intersects(Ray{{1, 2, 3}, {4, 5, 6}}, unit_box));
   BOOST_TEST(intersects(Ray{{1, 2, 3}, {-1, -2, -3}}, unit_box));
+
+  // origin is on the box
+  BOOST_TEST(intersects(Ray{{1, .5, .5}, {-1, 0, 0}}, unit_box));
+  BOOST_TEST(intersects(Ray{{1, .5, .5}, {1, 0, 0}}, unit_box));
+  BOOST_TEST(intersects(Ray{{.5, 1, .5}, {0, -1, 0}}, unit_box));
+  BOOST_TEST(intersects(Ray{{.5, 1, .5}, {0, 1, 0}}, unit_box));
+  BOOST_TEST(intersects(Ray{{.5, .5, 1}, {0, 0, -1}}, unit_box));
+  BOOST_TEST(intersects(Ray{{.5, .5, 1}, {0, 0, 1}}, unit_box));
+
+  BOOST_TEST(intersects(Ray{{0, 0, .5}, {1, 1, 0}}, unit_box));
+  BOOST_TEST(intersects(Ray{{0, 0, .5}, {-1, -1, 0}}, unit_box));
+  BOOST_TEST(intersects(Ray{{1, 1, .5}, {-1, -1, 0}}, unit_box));
+  BOOST_TEST(intersects(Ray{{1, 1, .5}, {1, 1, 0}}, unit_box));
+
+  BOOST_TEST(intersects(Ray{{1, 1, 1}, {-1, -1, -1}}, unit_box));
+  BOOST_TEST(intersects(Ray{{1, 1, 1}, {1, 1, 1}}, unit_box));
+  BOOST_TEST(intersects(Ray{{0, 0, 0}, {1, 1, 1}}, unit_box));
+  BOOST_TEST(intersects(Ray{{0, 0, 0}, {-1, -1, -1}}, unit_box));
+
+  BOOST_TEST(intersects(Ray{{0, 0.5, 0}, {-1, -2, 0}}, unit_box));
+  BOOST_TEST(intersects(Ray{{0.5, 0, 0.5}, {2, 0, -1}}, unit_box));
+  BOOST_TEST_WARN(intersects(Ray{{0, 0, 0}, {0, 1, -1}}, unit_box));
```

Comment:
Why did you use the `BOOST_TEST_WARN` macro?

Reply:
This test fails on my machine.
According to our discussion, it should return true since the origin of the ray is at the corner of the box, but the algorithm might not be robust enough for such cases.
I am not sure using BOOST_TEST_WARN is the right way for this purpose.

Reply:
I think we have a fundamental problem. If a ray origin is in (0,0,0), and first coordinate of ray direction is 0, then first value of `min_max` will be `nan`. As `nan` always compares false, it will result in the final `min_max` being `nan` and returning false. So, it cannot detect any rays on that plane.

What's even more tricky is that it *will* detect any rays on (*, 0, *) plane, as the first comparison will not be none. So it becomes finicky that it will intersect with some bounds, but not with others.

I think we may need to add some `is_nan()` checks for `tmin`, `tmax`, `min_max` and `max_min` to get a consistent behavior.

Reply:
Actually, I'm no longer sure what's going to happen if the ray origin is in min corner or max corner. Like, at all. We need a lot more testing for rays originating there: rays going inside, rays going outside, rays along the faces, rays along the edges, etc.

Reply:
An example:
```
intersects(Ray({1, 0, 0}, {-1, 0, 0}), unit_box) -> false
intersects(Ray({1, 0, 0}, { 0, 1, 0}), unit_box) -> true       
```

Reply:
"Actually, I'm no longer sure what's going to happen if the ray origin is in min corner or max corner." this alone should not be an issue, the issue is when you have a zero * inf. Is this what you mean? There are already many cases for what you described. Could you tell me what you found in detail for any case that doesn't have a zero*inf situation?  
  
Also, are we talking about improving the algorithm until it covers all these special issues (zero*infinity), or just add more tests to understand the issue more?

Reply:
I will add more tests to trigger the NaN for different cases, but I don't see the ray origin on the corners alone will cause any issue.